api/base/filters.py

```diff
@@ -20,53 +30,184 @@ def filter_queryset(self, request, queryset, view):
             return queryset.sort(*ordering)
         return queryset
 
-query_pattern = re.compile(r'filter\[\s*(?P<field>\S*)\s*\]\s*')
 
+class FilterMixin(object):
+    """ View mixin with helper functions for filtering. """
+
+    QUERY_PATTERN = re.compile(r'^filter\[(?P<field>\w+)\](\[(?P<op>\w+)\])?$')
 
-def query_params_to_fields(query_params):
-    return {
-        query_pattern.match(key).groupdict()['field']: value
-        for key, value in query_params.items()
-        if query_pattern.match(key)
+    MATCH_OPERATORS = ('contains', 'icontains')
+    MATCHABLE_FIELDS = (ser.CharField, ser.ListField)
+
+    DEFAULT_OPERATOR = 'eq'
+    DEFAULT_OPERATOR_OVERRIDES = {
+        ser.CharField: 'icontains',
+        ser.ListField: 'contains',
     }
 
+    NUMERIC_FIELDS = (ser.IntegerField, ser.DecimalField, ser.FloatField)
 
-class FilterMixin(object):
-    """ View mixin with helper functions for filtering. """
+    DATE_FIELDS = (ser.DateTimeField, ser.DateField)
+    DATETIME_PATTERN = re.compile(r'^\d{4}\-\d{2}\-\d{2}(?P<time>T\d{2}:\d{2}(:\d{2}(:\d{1,6})?)?)$')
 
-    DEFAULT_OPERATOR = 'eq'
+    COMPARISON_OPERATORS = ('gt', 'gte', 'lt', 'lte')
+    COMPARABLE_FIELDS = NUMERIC_FIELDS + DATE_FIELDS
+
+    LIST_FIELDS = (ser.ListField, )
 
     def __init__(self, *args, **kwargs):
         super(FilterMixin, self).__init__(*args, **kwargs)
         if not self.serializer_class:
             raise NotImplementedError()
 
-    def is_filterable_field(self, key):
-        try:
-            return key.strip() in self.serializer_class.filterable_fields
-        except AttributeError:
-            return key.strip() in self.serializer_class._declared_fields
+    @staticmethod
+    def _isinstance_any(instance, bases):
+        return any([isinstance(instance, base) for base in bases])
```

Comment:
The second argument to `isinstance` accepts a tuple, and isinstance returns True if the object is any one of the types.
https://docs.python.org/2/library/functions.html#isinstance

This function can be eliminated, and the many other uses of isinstance throughout can be written much more cleanly as:
`isinstance(instance, bases)`

(See `convert_value` for additional isinstance uses that can be simplified but don't use this function)